### PR TITLE
Added e2e for egress firewall

### PR DESCRIPTION
**Pull request by nerdalert**
- this commit adds some e2e testing for egress firewall tests
- tests include traffic explicitly allowed by the firewall
  crd config as well as traffic implicitly denied by the policy
- changed the wait.PollImmediate to look for a pods ip annotation

Signed-off-by: Brent Salisbury <bsalisbu@redhat.com>

**- What this PR does and why is it needed**

- commit adds e2e tests for the egress firewall feature

**- Special notes for reviewers**

- I used some external sites for the testing. Before I got too deep into more tests I wanted to get consensus on the basics. I will add more scenarios such as:
- more acl protocol combinations 
- testing updates/deletes to the policy
- fixed the `wait.PollImmediate` function to work. Will patch the other tests once this gets reviewed.

This tests only one policy per namespace as that is what is currently supported. Ty.

**- How to verify it**

PR can be verified with the following:
```
# Install kind v7
curl -Lo ./kind "https://github.com/kubernetes-sigs/kind/releases/download/v0.7.0/kind-$(uname)-amd64"
chmod +x ./kind
mv ./kind /some-dir-in-your-PATH/kind

# Install kind v8
curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.8.1/kind-$(uname)-amd64
chmod +x ./kind
mv ./kind /some-dir-in-your-PATH/kind

# Start ovn-kube kind
pushd $GOPATH/src/github.com/ovn-org/ovn-kubernetes/contrib
./kind.sh
# From the Kubernetes repo run 
pushd $GOPATH/src/k8s.io/kubernetes
kubetest --provider=local --deployment=kind --kind-cluster-name=ovn --test
```

### CODE REVIEW

**Comment by aojea** [comment]
> * I used some external sites for the testing. Before I got too deep into more tests I wanted to get consensus on the basics. I will add more scenarios such as:

adding real/external sites have some tradeoffs: flakes due to the connectivity from github infra, no IPv6, can not run without internetn,

You can create fake external host using docker, see this PR as an example

https://github.com/ovn-org/ovn-kubernetes/pull/1561/files#diff-6413e1dbafd6dd63b5d907ea3ae6f6f9R245-R272

File: test/e2e/e2e_test.go
Comment (aojea) [inline test/e2e/e2e_test.go:1229]: nit ginkgo.BeforeEach

File: test/e2e/e2e_test.go
Comment (aojea) [inline test/e2e/e2e_test.go:1327]: you can use this helper function, I think that there is no need to validate the IP, it's already validated in k8s
https://github.com/ovn-org/ovn-kubernetes/pull/1561/files#diff-6413e1dbafd6dd63b5d907ea3ae6f6f9R155-R171

File: test/e2e/e2e_test.go
Comment (aojea) [inline test/e2e/e2e_test.go:1350]: check if you can reuse some of the already code in the framework
https://github.com/kubernetes/kubernetes/blob/1c548c328a32c5b6819195760d1abba115ab4969/test/e2e/network/network_policy.go#L75

File: test/e2e/e2e_test.go
Comment (aojea) [inline test/e2e/e2e_test.go:1305]: You can use this snipped so you don't care about the deployment mode
https://github.com/ovn-org/ovn-kubernetes/pull/1561/files#diff-6413e1dbafd6dd63b5d907ea3ae6f6f9R384-R403

File: test/e2e/e2e_test.go
Comment (nerdalert) [reply test/e2e/e2e_test.go:1327]: The need for the `wait.PollImmediate` is because framework test pods are returning from `e2epod.WaitForPodNotPending` before they have an IP and the tests will fail/flake most of the time since there isn't an IP assigned yet. More detailed description [here](https://github.com/ovn-org/ovn-kubernetes/pull/1612#discussion_r471716294) on why I changed it to look for the IP annotation. Ty!

File: test/e2e/e2e_test.go
Comment (nerdalert) [reply test/e2e/e2e_test.go:1229]: Thanks, I can never keep up with whether dot imports are in or out of fashion :)

File: test/e2e/e2e_test.go
Comment (nerdalert) [reply test/e2e/e2e_test.go:1350]: Great point, I went through those and under the hood are spinning up a pod per each connect attempt with netcat. This is just using a single pod for each test to save time. I do think we can break it out into some functions to tighten it up. Will do it on a future PR and refactor some of the other tests unless you think its a show stopper here. Ty!

File: test/e2e/e2e_test.go
Comment (nerdalert) [reply test/e2e/e2e_test.go]: Nice, thanks and done! I hadn't seen this functionality, I like this a lot.

**Comment by nerdalert** [comment]
> > * I used some external sites for the testing. Before I got too deep into more tests I wanted to get consensus on the basics. I will add more scenarios such as:
> 
> adding real/external sites have some tradeoffs: flakes due to the connectivity from github infra, no IPv6, can not run without internetn,
> 
> You can create fake external host using docker, see this PR as an example
> 
> https://github.com/ovn-org/ovn-kubernetes/pull/1561/files#diff-6413e1dbafd6dd63b5d907ea3ae6f6f9R245-R272

@aojea I tend to agree even though those hosts are hardly ever not reachable (also enables offline testing in theory). I did mock external containers for the gateway tests. Will need to add the container IPs into the CRD definition which will be dynamic. 

Looking at your PR it is bumping KIND which changes the bridge to a named bridge and you have some helpers for getting container addresses so I think we either:
A) go ahead with the external IPs then I will convert them once your PR merges.
_or_
B) put this PR on hold and rebase once your PR goes in with the changes. 
Ty

**Comment by aojea** [comment]
Let's go with A then, no need to block 

> **A)** go ahead with the external IPs then I will convert them once your PR merges.

/lgtm

**Comment by JacobTanenbaum** [comment]
the test failures dont seem to have anything to do with the changes made in ths PR. 

lgtm

**Review by trozet — APPROVED** [review]
/lgtm just rebase please

**Pull request merged**
